### PR TITLE
refactor: Debug feature refactor

### PR DESCRIPTION
**Pull request by Evgenia-bit**
## Changes
- refactor debug screen
- refactor uikit screen
- fixed tests

## Checklist for self-check
- [x] Commits and PRs have been filed according to [the rules on the project](https://github.com/surfstudio/surf-flutter-app-template#workflow-in-a-repository).
- [x] The author is marked as an assigne and assigned mandatory reviewers.
- [x] Required labels marked
- [ ] Specified related tasks and/or related PRs.
- [x] Specified Changes.
- [ ] Attached videos/screenshots demonstrating the fix/feature.
- [x] All unspecified fields in the PR description deleted.
- [ ] New code covered by tests.

## Checklist for reviewers
- [ ] CI passed successfully _(with a green check mark)_.
- [ ] PR is atomic, by volume no more than 400 (+-) corrected lines (not including codogen).

Design:
- [ ] System design corresponds to the agreements on structure and architecture on the project.
- [ ] The code is decomposed into necessary and sufficient components.

Functionality:
- [ ] The code solves the problem.
- [ ] Any changes to the user interface are reasonable and look good.

Complexity:
- [ ] The code is clear, easy to read, functions are small, no more than 50 lines.
- [ ] The logic is not overcomplicated, there is no overengineering (no code sections that may be needed in the future, but no one knows about it).

Tests:
- [ ] Updated or added tests for mandatory components.
- [ ] The tests are correct, helpful, and well designed/developed.

Naming:
- [ ] The naming of variables, methods, classes and other components is understandable.

Comments:
- [ ] The comments are understandable and helpful.

Documentation:
- [ ] All labels are correct
- [ ] Technical documentation updated (after approval, updates last reviewer).

### CODE REVIEW

File: lib/features/debug/data/repositories/debug_repository.dart [hunk 1]
```diff
@@ -21,43 +22,24 @@ final class DebugRepository implements IDebugRepository {
...
-  }
-
-  @override
-  RequestOperation<void> saveProxyUrl(String proxyUrl) async {
+  RequestOperation<void> saveProxyUrl(String? proxyUrl) async {
     try {
       await _configStorage.setProxyUrl(proxyUrl: proxyUrl);
       return const ResultOk(null);
+    } on DioException catch (e, s) {
+      return Result.failed(Failure(original: e, trace: s));
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/data/repositories/debug_repository.dart:30]: ConfigStorage can't throw DioException. Let's leave just one catch.

File: lib/features/debug/data/repositories/debug_repository.dart [hunk 2]
```diff
@@ -21,43 +22,24 @@ final class DebugRepository implements IDebugRepository {
...
   }
 
   @override
-  RequestOperation<void> saveUrl(Url url) async {
+  RequestOperation<void> saveServerUrl(Url url) async {
     try {
       await _configStorage.setUrlType(urlType: _urlConverter.convertReverse(url));
       return const ResultOk(null);
+    } on DioException catch (e, s) {
+      return Result.failed(Failure(original: e, trace: s));
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/data/repositories/debug_repository.dart:42]: Same as above.

File: lib/features/debug/presentation/screens/debug/debug_model.dart [hunk 3]
```diff
@@ -1,62 +1,50 @@
...
   }
 
-  @override
-  void dispose() {
-    _themeService.removeListener(_updateTheme);
-  }
+  late final ValueNotifier<Url> _serverUrlState;
+
+  /// Current server url state.
+  ValueListenable<Url> get serverUrlState => _serverUrlState;
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/debug/debug_model.dart:25]: As far as I can see it is pure UI state. So it should be in `wm`, not in `model`.

File: lib/features/debug/presentation/screens/debug/debug_model.dart [hunk 4]
```diff
@@ -1,62 +1,50 @@
...
+  Future<void> saveProxyUrl(String url) async {
+    final newProxyUrl = url.isEmpty ? null : url;
+    await makeCall(() => _debugRepository.saveProxyUrl(newProxyUrl));
   }
 
-  /// Set theme mode for app.
-  void setThemeMode(ThemeMode themeMode) {
-    _themeService.updateThemeMode(themeMode);
+  /// Set value for serverUrlState.
+  // TODO(Evgenia-bit): убрать игнор после мержа ветки SNP-857-api в main
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/debug/debug_model.dart:39]: Leave comments in English, please.

File: lib/features/debug/presentation/screens/debug/debug_wm.dart [hunk 5]
```diff
@@ -1,162 +1,123 @@
...
+  void _showReloadAppSnackBar() {
+    _scaffoldMessenger.showSnackBar(SnackBar(content: Text(context.l10n.debugScreenReloadAppMessage)));
   }
 
   @override
-  void setThemeMode(ThemeMode? themeMode) {
+  void onSetThemeMode(ThemeMode? themeMode) {
     if (themeMode == null) return;
-    model.setThemeMode(themeMode);
+    // TODO(Evgenia-bit): добавится изменение темы, когда ветка с рефакторингом сервиса темы будет смержена.
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/debug/debug_wm.dart:118]: Leave comments in English, please.

File: lib/features/debug/presentation/screens/debug/debug_wm.dart [hunk 6]
```diff
@@ -1,162 +1,123 @@
...
-  /// Class that coordinates navigation for the whole app.
-  final AppRouter router;
+  final String? _initialProxyUrl;
 
-  late final _textEditingController = TextEditingController(text: model.proxyUrl);
+  late final TextEditingController _textEditingController;
 
-  @override
-  late final ValueNotifier<Url> urlState;
+  late final ScaffoldMessengerState _scaffoldMessenger;
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/debug/debug_wm.dart:64]: I don't think that we need `late final` variable. We can inject `ScaffoldMessengerState` through `DebugScreenWM` constructor.

File: lib/features/debug/presentation/screens/debug/debug_wm.dart [hunk 7]
```diff
@@ -1,162 +1,123 @@
...
-
-  /// Change proxyUrl value.
-  void setProxy() {}
-
-  /// Set theme mode for app.
-  void setThemeMode(ThemeMode? themeMode) {}
-
-  /// Navigate to ui kit screen.
-  void openUiKit();
+  void openUiKit() => _router.push(UiKitRouter());
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/debug/debug_wm.dart:122]: You can add `const` constructor here.

File: lib/features/debug/presentation/screens/ui_kit/ui_kit_screen.dart
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/ui_kit/ui_kit_screen.dart:1]: Let's also add `UIKitFlow`. Without scope, just `UIKitFlow`.

File: lib/features/debug/presentation/screens/debug/debug_screen.dart
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/debug/debug_screen.dart:1]: Let's also add `DebugFlow`. Without scope, just `DebugFlow`.

File: lib/features/debug/presentation/screens/ui_kit/ui_kit_wm.dart [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+UiKitWM uiKitScreenWMFactory(BuildContext context) {
+  final appScope = context.read<IAppScope>();
+
+  final model = UiKitModel(logWriter: appScope.logger);
+
+  return UiKitWM(model);
+}
+
+/// Interface for [UiKitWM].
+abstract class IUiKitWM with ILocalizationMixin, ThemeIModelMixin implements IWidgetModel {
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/ui_kit/ui_kit_wm.dart:20]:
```suggestion
abstract interface class IUiKitWM with ILocalizationMixin, ThemeIModelMixin implements IWidgetModel {
```

File: lib/features/debug/presentation/screens/debug/debug_wm.dart [hunk 9]
```diff
@@ -1,162 +1,123 @@
...
     logWriter: appScope.logger,
+    serverUrl: appConfig.url,
   );
 
-  return DebugScreenWidgetModel(model, appRouter);
+  return DebugScreenWM(model, router: appRouter, proxyUrl: appConfig.proxyUrl);
+}
+
+/// Interface for [DebugScreenWM].
+abstract class IDebugScreenWM with ILocalizationMixin implements IWidgetModel {
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/debug/debug_wm.dart:31]:
```suggestion
abstract interface class IDebugScreenWM with ILocalizationMixin implements IWidgetModel {
```

File: lib/features/debug/presentation/screens/ui_kit/ui_kit_wm.dart [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+  /// Callback of pressing the positive snack button.
+  void onPositiveSnackButtonPressed();
+}
+
+/// {@template ui_kit_widget_model.class}
+/// [WidgetModel] for [UiKitScreen].
+/// {@endtemplate}
+class UiKitWM extends WidgetModel<UiKitScreen, UiKitModel> with LocalizationMixin, ThemeWMMixin implements IUiKitWM {
+  late final ScaffoldMessengerState _scaffoldMessenger;
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/ui_kit/ui_kit_wm.dart:47]: I don't think that we need `late final` variable. We can inject `ScaffoldMessengerState` through `UiKitWM` constructor.

File: lib/features/debug/presentation/screens/ui_kit/ui_kit_wm.dart [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  UiKitWM(super._model);
+
+  @override
+  void initWidgetModel() {
+    _scaffoldMessenger = ScaffoldMessenger.of(context);
+    super.initWidgetModel();
+  }
+
+  @override
+  void switchTheme() => context.read<IAppScope>().themeService.switchTheme();
```
Comment (dzmitry-struk-surf) [inline lib/features/debug/presentation/screens/ui_kit/ui_kit_wm.dart:59]: You should change it after main branch merging.

File: lib/features/navigation/service/router.dart [hunk 12]
```diff
@@ -13,7 +16,7 @@ part 'router.gr.dart';
 /// Main point of the application navigation.
 /// {@endtemplate}
 @AutoRouterConfig(
-  replaceInRouteName: 'Flow,Route',
+  replaceInRouteName: 'Flow|Screen,Route',
```
Comment (dzmitry-struk-surf) [inline lib/features/navigation/service/router.dart:19]: You should leave just `Flow` here and add DebugFlow and UiKitFlow as I mentioned above.

File: lib/persistence/storage/config_storage/config_storage.dart
Comment (dzmitry-struk-surf) [inline lib/persistence/storage/config_storage/config_storage.dart:1]: I suggest to make `getProxyUrl` and `getUrlType` synchronous functions. What do you think?

**Review by dzmitry-struk-surf — CHANGES_REQUESTED** [review]
In general - nice refactoring.

File: lib/features/debug/presentation/screens/ui_kit/ui_kit_screen.dart
Comment (Evgenia-bit) [reply lib/features/debug/presentation/screens/ui_kit/ui_kit_screen.dart:1]: maybe it's worth fixing in the documentation that the screens have an entry point?

Comment (Evgenia-bit) [reply lib/features/debug/presentation/screens/ui_kit/ui_kit_screen.dart:1]: Screenshot 2024-03-04 at 11 51 13
currently, flow is added only for subfeature

Comment (dzmitry-struk-surf) [reply lib/features/debug/presentation/screens/ui_kit/ui_kit_screen.dart:1]: From my point of view, screen and subfeature - are quite the same, what do you think? I think that we should not use screen as it is now in docs, let's remove it and leave just subfeature.

File: lib/features/debug/presentation/screens/debug/debug_screen.dart
Comment (dzmitry-struk-surf) [reply lib/features/debug/presentation/screens/debug/debug_screen.dart:1]: You should move `DebugFlow` to the `debug` folder, it should not be  in the root.

**Review by Sadhorsephile — APPROVED** [review]

**Pull request merged**